skiko/src/jvmMain/kotlin/org/jetbrains/skiko/cleanup.kt

```diff
@@ -0,0 +1,47 @@
+@file:OptIn(ExperimentalCoroutinesApi::class)
+
+package org.jetbrains.skiko
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import java.nio.file.Path
+import java.time.Duration
+import java.time.Instant
+import kotlin.concurrent.thread
+import kotlin.io.path.*
+
+/**
+ * Will asynchronously cleanup 'stale' entries in the [SkikoProperties.dataPath]:
+ * If directories are older than [SkikoProperties.dataCleanupDays], then they will be deleted.
+ */
+internal fun launchSkikoDataDirCleanupIfNecessary(dataDir: Path) {
+    thread(
+        name = ".skiko cleanup",
+        priority = Thread.MIN_PRIORITY
+    ) {
+        try {
+            withDirectoryLock(dataDir) {
+                doCleanupSkikoDataDir()
+            }
+        } catch (t: Throwable) {
+            Logger.error(t) { "Exception occurred during .skiko cleanup" }
+        }
+    }
+}
+
+private fun doCleanupSkikoDataDir() {
+    val dir = Path(SkikoProperties.dataPath)
+
+    dir.listDirectoryEntries().forEach { entry ->
+        if (!entry.isDirectory()) return@forEach
```

Comment:
Does the parent folder has the same time (.skiko/hash/) as the file accessed in it .skiko/hash/skiko.so)? If not, it is better to iterate over the files.

Reply:
The lastAccessTime of the `.skiko/hash` content will be ensured to be updated; therefore, the simple cleaning rule would be: Only iterate over directories and clean up stuff which was not accessed